### PR TITLE
refactor to use strings , add expiry,  and convert from nanoMassa to massa in all Operations

### PR DESCRIPTION
**Pull request by B-Naoufal**

### CODE REVIEW

**Review by Thykof — APPROVED** [review]

**Review by mazmassa — APPROVED** [review]

File: wails-frontend/src/utils/massaFormat.tsx [hunk 1]
```diff
@@ -1,5 +1,7 @@
 import { fromMAS, toMAS } from '@massalabs/massa-web3';
 
+const MASSA_SCALING_FACTOR = 9;
```
Comment (mazmassa) [inline wails-frontend/src/utils/massaFormat.tsx:3]: Is that decimals!? if yes, why not:
```suggestion
const MASSA_DECIMALS = 9;
```

File: internal/handler/wallet/sign.go [hunk 2]
```diff
@@ -206,11 +208,11 @@ func (s *walletSign) prepareCallSCPromptRequest(msg *callsc.MessageContent,
...
 			Description:   description,
-			Fees:          fees,
+			Fees:          strconv.FormatUint(fees, 10),
 			OperationType: "Call SC",
 			OperationID:   msg.OperationID,
-			GasLimit:      msg.GasLimit,
-			Coins:         msg.Coins,
+			GasLimit:      strconv.FormatUint(msg.GasLimit, 10),
+			Coins:         strconv.FormatUint(msg.Coins, 10),
 			Address:       msg.Address,
```
Comment (peterjah) [inline internal/handler/wallet/sign.go:216]: expiry missing here?

File: wails-frontend/src/utils/massaFormat.tsx [hunk 3]
```diff
@@ -37,16 +43,22 @@ export function toNanoMASS(str: string): number {
...
+  const options = {
     minimumFractionDigits: 2,
     maximumFractionDigits,
-  });
+  };
+
+  // Format the number as a string with separators
+  const formattedNum = numInMas.toLocaleString(locale, options);
+
+  return formattedNum;
```
Comment (peterjah) [inline wails-frontend/src/utils/massaFormat.tsx:61]: why modify this code to do the same thing ? you can keep the comment but the previous version was more clean imho...

File: wails-frontend/src/pages/PasswordPromptHandler/BuySellRoll/BuySellRoll.tsx [hunk 4]
```diff
@@ -28,13 +28,13 @@ export function BuySellRoll(props: PromptRequestData) {
       <div className="flex w-full items-center justify-between">
         <p>{Intl.t('password-prompt.sign.coins')}</p>
         <p>
-          {formatStandard(Number(Coins))} {masToken}
+          {formatStandard(toMASString(Coins))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/BuySellRoll/BuySellRoll.tsx:31]: why not just `formatStandard(Coins, Unit.NanoMAS)` ???

File: wails-frontend/src/pages/PasswordPromptHandler/BuySellRoll/BuySellRoll.tsx [hunk 5]
```diff
@@ -28,13 +28,13 @@ export function BuySellRoll(props: PromptRequestData) {
...
         </p>
       </div>
       <div className="flex w-full items-center justify-between">
         <p>{Intl.t('password-prompt.sign.fees')}</p>
         <p>
-          {formatStandard(Number(Fees))} {masToken}
+          {formatStandard(toMASString(Fees))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/BuySellRoll/BuySellRoll.tsx:37]: formatStandard(Fees, Unit.NanoMAS)

File: wails-frontend/src/pages/PasswordPromptHandler/ExecuteSC.tsx/ExecuteSc.tsx [hunk 6]
```diff
@@ -25,13 +26,23 @@ export function ExecuteSC(props: PromptRequestData) {
       <div className="flex w-full items-center justify-between">
         <p>{Intl.t('password-prompt.sign.max-coins')} </p>
         <p>
-          {formatStandard(Number(MaxCoins))} {masToken}
+          {formatStandard(toMASString(MaxCoins))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/ExecuteSC.tsx/ExecuteSc.tsx:29]: formatStandard(MaxCoins, Unit.NanoMAS)

File: wails-frontend/src/pages/PasswordPromptHandler/ExecuteSC.tsx/ExecuteSc.tsx [hunk 7]
```diff
@@ -25,13 +26,23 @@ export function ExecuteSC(props: PromptRequestData) {
...
+      </div>
+      <div className="flex w-full items-center justify-between">
+        <p>{Intl.t('password-prompt.sign.expiry')} </p>
+        <p>{Expiry}</p>
+      </div>
       <div className="flex w-full items-center justify-between">
         <p>{Intl.t('password-prompt.sign.fees')}</p>
         <p>
-          {formatStandard(Number(Fees))} {masToken}
+          {formatStandard(toMASString(Fees))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/ExecuteSC.tsx/ExecuteSc.tsx:45]: formatStandard(Fees, Unit.NanoMAS)

File: wails-frontend/src/pages/PasswordPromptHandler/SignSC/CallSc.tsx [hunk 8]
```diff
@@ -95,13 +95,13 @@ export function CallSc(props: PromptRequestData) {
         <div className="flex w-full items-center justify-between">
           <p>{Intl.t('password-prompt.sign.coins')}</p>
           <p>
-            {formatStandard(Number(Coins))} {masToken}
+            {formatStandard(toMASString(Coins))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/SignSC/CallSc.tsx:98]: formatStandard(Coins, Unit.NanoMAS)

File: wails-frontend/src/pages/PasswordPromptHandler/SignSC/CallSc.tsx [hunk 9]
```diff
@@ -95,13 +95,13 @@ export function CallSc(props: PromptRequestData) {
...
           </p>
         </div>
         <div className="flex w-full items-center justify-between">
           <p>{Intl.t('password-prompt.sign.fees')}</p>
           <p>
-            {formatStandard(Number(Fees))} {masToken}
+            {formatStandard(toMASString(Fees))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/SignSC/CallSc.tsx:104]: formatStandard(Fees, Unit.NanoMAS)

File: wails-frontend/src/pages/PasswordPromptHandler/Transaction/Transaction.tsx [hunk 10]
```diff
@@ -46,15 +46,15 @@ export function Transaction(props: PromptRequestData) {
         <div className="flex w-full items-center justify-between">
           <p>{Intl.t('password-prompt.sign.sending-amount')}</p>
           <p>
-            {formatStandard(Number(Amount))} {masToken}
+            {formatStandard(toMASString(Amount))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/Transaction/Transaction.tsx:49]: +1

File: wails-frontend/src/pages/PasswordPromptHandler/Transaction/Transaction.tsx [hunk 11]
```diff
@@ -46,15 +46,15 @@ export function Transaction(props: PromptRequestData) {
...
           </p>
         </div>
       </div>
 
       <div className="flex w-full items-center justify-between">
         <p>{Intl.t('password-prompt.sign.fees')}</p>
         <p>
-          {formatStandard(Number(Fees))} {masToken}
+          {formatStandard(toMASString(Fees))} {masToken}
```
Comment (peterjah) [inline wails-frontend/src/pages/PasswordPromptHandler/Transaction/Transaction.tsx:57]: +1

File: wails-frontend/src/utils/massaFormat.tsx [hunk 12]
```diff
@@ -20,6 +22,10 @@ export function toMASS(num: string | number): number {
   return Number(toMAS(num));
 }
 
+export function toMASString(num: string): string {
+  return toMAS(num).toFixed(MASSA_SCALING_FACTOR).toString().replace(/\.?0+$/, '');
+}
+
 export function toNanoMASS(str: string): number {
   const formattedString = str?.replace(/[^0-9.-]/g, ''); // Remove non-numeric characters
```
Comment (peterjah) [inline wails-frontend/src/utils/massaFormat.tsx:30]: wtf is this ? :'(

Comment (peterjah) [reply wails-frontend/src/utils/massaFormat.tsx:30]: i propose you tackle this in another PR...

**Review by peterjah — CHANGES_REQUESTED** [review]

File: wails-frontend/src/utils/massaFormat.tsx [hunk 13]
```diff
@@ -20,6 +22,10 @@ export function toMASS(num: string | number): number {
   return Number(toMAS(num));
 }
 
+export function toMASString(num: string): string {
+  return toMAS(num).toFixed(MASSA_SCALING_FACTOR).toString().replace(/\.?0+$/, '');
+}
+
```
Comment (peterjah) [inline wails-frontend/src/utils/massaFormat.tsx:28]: to remove.
Remove the trailing zeros in the formatStandar function

**Review by peterjah — CHANGES_REQUESTED** [review]

**Review by peterjah — APPROVED** [review]

File: wails-frontend/src/utils/massaFormat.tsx (hunk 12 above)
Comment (B-Naoufal) [reply wails-frontend/src/utils/massaFormat.tsx]: toMASS, and toNanoMASS will be removed in another PR

**Comment by longgamma-massa** [comment]
What to Q/A and how @B-Naoufal ?

**Comment by Thykof** [comment]
> What to Q/A and how @B-Naoufal ?

it's lot of refacto so you can do some non regression test, also from a feature/bugfix point of view @B-Naoufal or @peterjah could you explain how to QA or QA yourself?

**Pull request merged**